### PR TITLE
refactor: migrate project to multi-module

### PR DESCRIPTION
**Pull request by zrock-dev**
#### Resolved task/US: ([CS-49: Configure the SonarQube to the subproject ASTProcessor](https://app.clickup.com/t/868avu2a0))
* * * *

## What's in this pull request?

1. Migrated src and ASTProcessor into a multi-module project respectively. They share logic using[ coventional plugins](https://docs.gradle.org/current/userguide/sharing_build_logic_between_subprojects.html#sec:sharing_logic_via_convention_plugins) as Gradle recommends.
2. Configured project to have an [aggregated coverage report](https://docs.gradle.org/6.4-rc-1/samples/sample_jvm_multi_project_with_code_coverage.html). This way SonarQube will perform an analisis of all the projects.

## Type of change

- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## How Has This Been Tested?

1. Build the project to ensure no tests are brokern `$ ./gradlew build`
2. Execute local sonar analisis to ensure the complete coverage `$ ./sonar_analysis-v1.0.sh*`

## Checklist:
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have checked my code and corrected any misspellings

- - -
This change is [Reviewable](https://reviewable.io/reviews/SoftDevSix/Argos/40)

### CODE REVIEW

File: app/build.gradle.kts [hunk 1]
```diff
@@ -0,0 +1,17 @@
+plugins {
+    id("buildlogic.java-application-conventions")
```
Comment (salet-gutierrez-jalafund) [inline app/build.gradle.kts:2]: you could add this dependency inside `libs.versions.toml`

File: build-logic/settings.gradle.kts [hunk 2]
```diff
@@ -0,0 +1,15 @@
+/*
+ * This file was generated by the Gradle 'init' task.
+ *
+ * This settings file is used to specify which projects to include in your build-logic build.
+ * This project uses @Incubating APIs which are subject to change.
+ */
+
+dependencyResolutionManagement {
+    // Reuse version catalog from the main build.
```
Comment (salet-gutierrez-jalafund) [inline build-logic/settings.gradle.kts:2]: Is this comment necessary? I think the comment at the beginning would be enough

File: build-logic/src/main/kotlin/buildlogic.java-common-conventions.gradle.kts [hunk 3]
```diff
@@ -0,0 +1,42 @@
+plugins {
+    // Apply the java Plugin to add support for Java.
+    java
+    `jacoco`
+    id("org.sonarqube")
```
Comment (salet-gutierrez-jalafund) [inline build-logic/src/main/kotlin/buildlogic.java-common-conventions.gradle.kts:4]: I don't know if the dependencies that have their versions can be put inside the `libs.versions.toml` file, if not then I would like you to tell me the reason please.

File: build-logic/src/main/kotlin/buildlogic.java-common-conventions.gradle.kts [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+}
+
+repositories {
+    // Use Maven Central for resolving dependencies.
```
Comment (salet-gutierrez-jalafund) [inline build-logic/src/main/kotlin/buildlogic.java-common-conventions.gradle.kts]: Delete these comments, the ones you think are not necessary, from the other files as well if you think it is convenient.

**Review by salet-gutierrez-jalafund — CHANGES_REQUESTED** [review]

File: app/build.gradle.kts (hunk 1 above)
Comment (zrock-dev) [reply app/build.gradle.kts:2]: The buildlogic.java-application-conventions is part of the project and doesn't have version. It can be found at build-logic/src/main/kotlin/buildlogic.java-common-conventions.gradle.kts

File: build-logic/settings.gradle.kts (hunk 2 above)
Comment (zrock-dev) [reply build-logic/settings.gradle.kts:2]: Yes as it is to inform the next line is about depedency versioning.

File: build-logic/src/main/kotlin/buildlogic.java-common-conventions.gradle.kts (hunk 4 above)
Comment (zrock-dev) [reply build-logic/src/main/kotlin/buildlogic.java-common-conventions.gradle.kts]: Done

File: build-logic/src/main/kotlin/buildlogic.java-common-conventions.gradle.kts (hunk 3 above)
Comment (zrock-dev) [reply build-logic/src/main/kotlin/buildlogic.java-common-conventions.gradle.kts:4]: The version of this plugin comes from `libs.versions.toml` which I have updated to use the `libs.versions.toml `` file

File: code-analyzer/src/main/java/edu/usb/argos/ASTProcessor/antlr/JavaLexer.interp
Comment (salet-gutierrez-jalafund) [inline code-analyzer/src/main/java/edu/usb/argos/ASTProcessor/antlr/JavaLexer.interp:1]: why did you enter this file, if I'm not mistaken, if it is generated by the .g4 file it should not allow these files to be uploaded, if not all the classes generated by the .g4 file should be locally in our magic and those classes will be seen when doing the build, it would be good if you deleted it and only this file is generated when doing the build.

File: code-analyzer/src/main/java/edu/usb/argos/ASTProcessor/antlr/JavaLexer.java
Comment (salet-gutierrez-jalafund) [inline code-analyzer/src/main/java/edu/usb/argos/ASTProcessor/antlr/JavaLexer.java:1]: same here

File: code-analyzer/src/main/java/edu/usb/argos/ASTProcessor/antlr/JavaLexer.tokens
Comment (salet-gutierrez-jalafund) [inline code-analyzer/src/main/java/edu/usb/argos/ASTProcessor/antlr/JavaLexer.tokens:1]: same here

**Review by salet-gutierrez-jalafund — CHANGES_REQUESTED** [review]

File: build-logic/settings.gradle.kts (hunk 2 above)
Comment (salet-gutierrez-jalafund) [reply build-logic/settings.gradle.kts:2]: Okay, then it would be good if you left the comment.

File: build.gradle.kts [hunk 5]
```diff
@@ -1,55 +1,41 @@
...
-
-application {
-	mainClass.set("com.softdevsix.argos.ArgosApplication")
+    mavenCentral()
 }
 
-tasks.withType<Test> {
-	useJUnitPlatform()
-}
+// tag::coverageTask[]
```
Comment (salet-gutierrez-jalafund) [inline build.gradle.kts:11]: remove this comment

File: build.gradle.kts [hunk 6]
```diff
@@ -1,55 +1,41 @@
...
-		xml.required = true
-		csv.required = false
-		html.required = true
-	}
+    reports {
+        xml.required = true
+        html.required = false
+    }
 }
+// end::coverageTask[]
```
Comment (salet-gutierrez-jalafund) [inline build.gradle.kts:29]: remove this comment

File: settings.gradle.kts [hunk 7]
```diff
@@ -1,2 +1,7 @@
+pluginManagement {
+    includeBuild("build-logic")
+}
+
 rootProject.name = "Argos"
-include(":ASTProcessor")
```
Comment (salet-gutierrez-jalafund) [inline settings.gradle.kts:2]: Why did you remove `ASTProcessor`? Since it is still in the project, would it be useful to have `ASTProcessor` as normal directories or why do you still keep it in the folders but remove it here?

**Review by salet-gutierrez-jalafund — CHANGES_REQUESTED** [review]

File: build.gradle.kts (hunk 5 above)
Comment (zrock-dev) [reply build.gradle.kts:11]: The semantic of this specific comment is used by Gradle to tag the created task

File: build.gradle.kts (hunk 6 above)
Comment (zrock-dev) [reply build.gradle.kts:29]: The semantic of this specific comment is used by Gradle to tag the created task

File: settings.gradle.kts (hunk 7 above)
Comment (zrock-dev) [reply settings.gradle.kts:2]: I have moved AST Processor inside the code-analyzer module. This was done following our architecture guidelines.

File: code-analyzer/src/main/java/edu/usb/argos/ASTProcessor/antlr/JavaLexer.interp
Comment (zrock-dev) [reply code-analyzer/src/main/java/edu/usb/argos/ASTProcessor/antlr/JavaLexer.interp:1]: I understand. I have deleted that file and updated .gitignore for git to ignore that path

File: code-analyzer/src/main/java/edu/usb/argos/ASTProcessor/antlr/JavaLexer.java
Comment (zrock-dev) [reply code-analyzer/src/main/java/edu/usb/argos/ASTProcessor/antlr/JavaLexer.java:1]: I understand. I have deleted that file and updated .gitignore for git to ignore that path

File: code-analyzer/src/main/java/edu/usb/argos/ASTProcessor/antlr/JavaLexer.tokens
Comment (zrock-dev) [reply code-analyzer/src/main/java/edu/usb/argos/ASTProcessor/antlr/JavaLexer.tokens:1]: I understand. I have deleted that file and updated .gitignore for git to ignore that path

**Review by zrock-dev — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/SoftDevSix/Argos/40)* status: 0 of 38 files reviewed, 10 unresolved discussions (waiting on @salet-gutierrez-jalafund and @samuel-escalera-jalafund)

**Comment by zrock-dev** [comment]
This PR depends on #38 to have its conversations solved.

**Review by zamoranoalicia — DISMISSED** [review]
LGTM

**Comment by zrock-dev** [comment]
SonarQube has rejected this PR because of the "Lines of code" rules.
image

I have created a request on the IT requests channel
image
https://teams.microsoft.com/l/message/19:1d8ec1be4a274712b3fc71413121cb27@thread.tacv2/1732533765222?tenantId=1f9b1311-c757-4228-b0e2-7d4ade7932aa&groupId=ea88865f-3731-4cde-b04d-f8e5546fcaf7&parentMessageId=1732215475904&teamName=Progra%206%20-%20Software%20Development&channelName=IT%20Requests&createdTime=1732533765222

**Review by salet-gutierrez-jalafund — APPROVED** [review]

**Review by samuel-escalera-jalafund — APPROVED** [review]

File: app/src/main/java/edu/usb/argos/service/RulesValidator.java [hunk 8]
```diff
@@ -1,9 +1,9 @@
...
+package edu.usb.argos.service;
 
-import com.softdevsix.argos.domain.CodeComplexityRules;
-import com.softdevsix.argos.domain.Rules;
-import com.softdevsix.argos.domain.RulesRequestMap;
-import com.softdevsix.argos.exception.*;
+import edu.usb.argos.domain.CodeComplexityRules;
+import edu.usb.argos.domain.Rules;
+import edu.usb.argos.domain.RulesRequestMap;
+import edu.usb.argos.exception.*;
```
Comment (zamoranoalicia) [inline app/src/main/java/edu/usb/argos/service/RulesValidator.java:6]: please dont use this kind of imports

**Review by zamoranoalicia — APPROVED** [review]
LGTM

File: app/src/main/java/edu/usb/argos/service/RulesValidator.java (hunk 8 above)
Comment (zamoranoalicia) [reply app/src/main/java/edu/usb/argos/service/RulesValidator.java:6]: Since this changes are needed we can address this changes on the next code changes
